base/multidimensional.jl

```diff
@@ -401,35 +401,38 @@ module IteratorsMD
         iterfirst, iterfirst
     end
     @inline function iterate(iter::CartesianIndices, state)
-        valid, I = __inc(state.I, iter.indices)
+        valid, I = __inc(state.I, iter.indices, Val(ndims(iter)))
         valid || return nothing
         return CartesianIndex(I...), CartesianIndex(I...)
     end
 
     # increment & carry
     @inline function inc(state, indices)
-        _, I = __inc(state, indices)
+        _, I = __inc(state, indices, Val(length(state)))
         return CartesianIndex(I...)
     end
 
     # Unlike ordinary ranges, CartesianIndices continues the iteration in the next column when the
     # current column is consumed. The implementation is written recursively to achieve this.
     # `iterate` returns `Union{Nothing, Tuple}`, we explicitly pass a `valid` flag to eliminate
     # the type instability inside the core `__inc` logic, and this gives better runtime performance.
-    __inc(::Tuple{}, ::Tuple{}) = false, ()
-    @inline function __inc(state::Tuple{Int}, indices::Tuple{OrdinalRangeInt})
+    __inc(::Tuple{}, ::Tuple{}, ::Val) = false, ()
```

Comment:
> It's also called at R435 as the last input ndim is passed deeper without any change.

A nice solution! Let's add one more comment for future you/me :)

```suggestion
    # The third argument ndim is used to preserve the original dimension information to elliminate
    # unnecessary boundary checks and thus enabling vectorization -- this makes low dimensional
    # case, especially 1d and 2d, much happier.
    __inc(::Tuple{}, ::Tuple{}, ::Val) = false, ()
```